### ISSUE TITLE
Some rules can't be suppressed through .vnurc config file

### ISSUE THREAD

**Issue by nschonni** [validator/validator#1070]
From scanning the `mdn/content` I found that the following rules appear not to be suppressible
- `Bad value “.*” for attribute “href” on element “a”: Illegal character in fragment: “.*” is not allowed.`
- `Element “head” is missing a required instance of child element “title”.`
- `Non-space characters found without seeing a doctype first.`

Here is the config file used
```
# Add regex filters for the Nu Validator error messages for proposed changes
# not yet added to the validator.
#
# Ignore failures
# .*Bad value “.*” for attribute “href” on element “a”: Illegal character in fragment: “#” is not allowed\.\n
# .*Element “head” is missing a required instance of child element “title”\.\n
# .*Non-space characters found without seeing a doctype first\. Expected “<!DOCTYPE html>”\.\n
An “img” element must have an “alt” attribute, except under certain conditions. For details, consult guidance on providing text alternatives for images.
Bad value “.*” for attribute “href” on element “a”: Illegal character in fragment: “.*” is not allowed.
Duplicate ID “.*”.
Element “head” is missing a required instance of child element “title”.
Non-space characters found without seeing a doctype first.
```

In understand that scanning files with frontmatter is a non-goal for the tool, but I think the suppressions not working might actually be a bug

**Comment by sideshowbarker** [comment]
Yeah it surprises me that those aren’t getting filtered — but it’s nice to have an interesting problem to investigate, so I’ll make time to look into it.

> In understand that scanning files with frontmatter is a non-goal for the tool

Well — I meant to give you a heads-up about this before now — but there’s a relatively easy way you can check the rendered output of the entirety of MDN (rather the raw sources). Here’s how:

I’m pretty sure you already have an mdn/yari clone, so within your yari clone and with `CONTENT_ROOT` set to wherever your mdn/content/files directory is, do this:

1. `yarn build`
  That builds the rendered output of the entire tree. In my environment, it takes about 9 minutes to complete.

2. `java -jar ~/vnu.jar --skip-non-html --errors-only client/build/en-us/docs`
  In my environment, that takes about 4 minutes to complete, and reports 17846 errors.

**Comment by nschonni** [comment]
Yeah, I prefer the in-place VNU scanning, because I can click on the "URL"/path directly to go to the line in VS code 😉

**Comment by nschonni** [comment]
> An “img” element must have an “alt” attribute, except under certain conditions. For details, consult guidance on providing text alternatives for images.

This cleanup will probably take a lot of work

> Bad value “.*” for attribute “href” on element “a”: Illegal character in fragment: “.*” is not allowed.

This one is for those broken by design Matrix/Riot URLs

Other than that, I think with the other pending PRs, if this gets resolved then it's pretty close to being able to add it to mdn/content. https://github.com/mdn/content/pull/418 probably won't get accepted, but that's the branch I use as the base for running

**Comment by sideshowbarker** [comment]
> > An “img” element must have an “alt” attribute, except under certain conditions. For details, consult guidance on providing text alternatives for images.
> 
> This cleanup will probably take a lot of work

Yeah. If/when we ever might add some automated CI validation of the entire tree, I think we will need to filter all those out. But short of that, we could still have CI automation that runs on every new PR and checks just the changed files for that PR.

So in that case, we wouldn’t want to filter those errors — for any newly added pages, that check would ensure we don’t get any new images added without `alt` attributes, and for any existing pages that get touched in a PR, it would give us a chance o fix them.

> > Bad value “._” for attribute “href” on element “a”: Illegal character in fragment: “._” is not allowed.
> 
> This one is for those broken by design Matrix/Riot URLs

Yeah, so maybe we would need to add a filter for those out always. As far as I know, there’s no way o “fix” them — no alternative URLs.

But anyway, with those or any other necessary filters in place, it seems like it would be practical to have some kind of automation running under CI that for every PR, invokes the HTML checker on all changed files in that PR. It would literally take only seconds to run each time.

**Comment by nschonni** [comment]
And with a problem matcher running on the non-complied code, it can comment inline on the issues by registering a problem matcher like https://github.com/w3c/aria-practices/blob/master/.github/workflows/vnu-jar-problem-matcher.json

**Comment by nschonni** [comment]
`Non-space characters found without seeing a doctype first. Expected “<!DOCTYPE html>”.` is actually working, just double checked an my suppression was missing the end of the statement. Other 2 aren't though

**Comment by nschonni** [comment]
Just tried with `Element “head” .* “title”.` and it suppressed, but I didn't see any funny characters in https://github.com/validator/validator/blob/master/src/nu/validator/messages/MessageEmitterAdapter.java#L365

**Comment by nschonni** [comment]
OK, so `Element “head” is missing .*“title”.` seems to be the minimal wildcard to allow it to pass, anything beyond and it seems to stop filtering again

**Comment by sideshowbarker** [comment]
> OK, so `Element “head” is missing .*“title”.` seems to be the minimal wildcard to allow it to pass, anything beyond and it seems to stop filtering again

OK thanks much — I just now did some local debugging and I see that this is caused by the fact the internal error message that the checker sees is different than what it ends up reporting.

At the point where the checker tries to match the message, the message text is this:

> Element “head” is missing required child element “title”

That is, *“missing required child element”* vs *“missing a required instance of child element”*.

I remember running into this before, and think this is because of a difference between the genereal message strings that jing uses, and the custom messages the checker code constructs.

Anyway, I will figure out how to fix this so that it matches as expected — without you as a user needing to drop any part of the user-visible message string in the filter pattern.

**Comment by sideshowbarker** [comment]
I’ve not forgotten about this — just have not had time to get back to it yet. But it turns out that fixing this is pretty complicated. The reason is, the messages in these cases aren’t emitted as a single string in the code but instead through a series of calls. So without changing how that’s done in the source — which is a relatively big deal to change — we can’t match the string as a whole. But it’s possible to special case them — short of trying to make the right general fix — and I will try to make time to do that this week.

**Comment by nschonni** [comment]
If you do go down the route of a bigger error suppression refactor, I'd vote for error (and warning) codes/names a-la pylint/eslint/markdownlint. I think there is already an issue somewhere tracking/suggesting that though. E001, W005 ...

**Comment by pepelsbey** [comment]
Maybe I’m doing something wrong, but I also can’t suppress the following error:

I’m running

```sh
java -jar node_modules/vnu-jar/build/dist/vnu.jar --filterfile .vnurc index.html
```

…with `.vnurc`

```
Element “style” not allowed as child of element “body” in this context. (Suppressing further errors from this subtree.)
```

…for this file

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Title</title>
</head>
<body>
    <style>
        body {
            background-color: tomato;
        }
    </style>
</body>
</html>
```

**Comment by sideshowbarker** [comment]
Fixed in https://github.com/validator/validator/commit/1b0d4c3a7b488ac502494d06f94254ae3eea4f98. So, please try with the latest jar, and if you run into any problems, raise issues.
